File: backend/app/schemas/dataset.py
"""Pydantic-схемы для датасетов."""
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class DatasetPreview(BaseModel):
    columns: list[str]
    rows: list[list[Any]]
    dtypes: dict[str, str]


class DatasetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    original_filename: str
    file_size_bytes: int
    format: str
    n_rows: int | None
    n_cols: int | None
    uploaded_at: datetime


class DatasetWithPreview(DatasetResponse):
    preview: DatasetPreview


class DatasetUsageResponse(BaseModel):
    """
    Используется на странице удаления, чтобы показать пользователю, сколько
    артефактов уйдёт вместе с датасетом (см. .knowledge/architecture/api-contract.md,
    Спринт 6, Phase 4.2).

    `analyses_count` — все анализы датасета без фильтра по статусу.
    `reports_count` — только PDF-отчёты в статусе success (failed/pending для UI
    бесполезны: фронт спрашивает «сколько готовых артефактов потеряет
    пользователь»).
    """

    analyses_count: int
    reports_count: int

File: frontend/src/types/dataset.ts
// Зеркало backend/app/schemas/dataset.py.

export interface Dataset {
  id: string;
  original_filename: string;
  file_size_bytes: number;
  format: "csv" | "xlsx";
  n_rows: number | null;
  n_cols: number | null;
  uploaded_at: string;
}

export interface DatasetPreview {
  columns: string[];
  rows: Array<Array<string | number | boolean | null>>;
  dtypes: Record<string, string>;
}

export interface DatasetWithPreview extends Dataset {
  preview: DatasetPreview;
}

// GET /api/datasets/{id}/usage (Спринт 6, Phase 4.2): счётчики связанных
// артефактов для confirm-диалога удаления. reports_count считает только
// success-PDF (failed/pending исключены).
export interface DatasetUsage {
  analyses_count: number;
  reports_count: number;
}
